c-bindings/c-bindings.go

```diff
@@ -156,5 +160,236 @@ func jsonnet_tla_code(vmRef *C.struct_JsonnetVm, key, value *C.char) {
+	case float64:
+		*out = C.double(f)
+		return 1
+	case int:
+		*out = C.double(f)
+		return 1
+	}
+
+	return 0
+}
+
+//export jsonnet_json_extract_bool
+func jsonnet_json_extract_bool(vmRef *C.struct_JsonnetVm, json *C.struct_JsonnetJsonValue) C.int {
+	_ = getVM(vmRef) // Here we check it, is it ok?
+	v := getJSONValue(json)
+	b, ok := v.obj.(bool)
+
+	if !ok {
+		return 2
+	}
+
+	if b {
+		return 1
+	}
+
+	return 0
+}
+
+//export jsonnet_json_extract_null
+func jsonnet_json_extract_null(vmRef *C.struct_JsonnetVm, json *C.struct_JsonnetJsonValue) C.int {
+	_ = getVM(vmRef) // Here we check it, is it ok?
+	v := getJSONValue(json)
+
+	if v.obj == nil {
+		return 1
+	}
+
+	return 0
+}
+
+//export jsonnet_json_make_string
+func jsonnet_json_make_string(vmRef *C.struct_JsonnetVm, v *C.char) *C.struct_JsonnetJsonValue {
+	return createJSONValue(vmRef, C.GoString(v))
+}
+
+//export jsonnet_json_make_number
+func jsonnet_json_make_number(vmRef *C.struct_JsonnetVm, v C.double) *C.struct_JsonnetJsonValue {
+	return createJSONValue(vmRef, float64(v))
+}
+
+//export jsonnet_json_make_bool
+func jsonnet_json_make_bool(vmRef *C.struct_JsonnetVm, v C.int) *C.struct_JsonnetJsonValue {
+	return createJSONValue(vmRef, v != 0)
+}
+
+//export jsonnet_json_make_null
+func jsonnet_json_make_null(vmRef *C.struct_JsonnetVm) *C.struct_JsonnetJsonValue {
+	return createJSONValue(vmRef, nil)
+}
+
+//export jsonnet_json_make_array
+func jsonnet_json_make_array(vmRef *C.struct_JsonnetVm) *C.struct_JsonnetJsonValue {
+	return createJSONValue(vmRef, []interface{}{})
+}
+
+//export jsonnet_json_array_append
+func jsonnet_json_array_append(vmRef *C.struct_JsonnetVm, arr *C.struct_JsonnetJsonValue, v *C.struct_JsonnetJsonValue) {
+	_ = getVM(vmRef) // Here we check it, is it ok?
+	json := getJSONValue(arr)
+	slice, ok := json.obj.([]interface{})
+
+	if !ok {
+		fmt.Fprintf(os.Stderr, "array should be provided")
+		os.Exit(1)
+	}
+
+	val := getJSONValue(v)
+	json.obj = append(slice, val.obj)
+	json.children = append(json.children, v)
+}
+
+//export jsonnet_json_make_object
+func jsonnet_json_make_object(vmRef *C.struct_JsonnetVm) *C.struct_JsonnetJsonValue {
+	return createJSONValue(vmRef, make(map[string]interface{}))
+}
+
+//export jsonnet_json_object_append
+func jsonnet_json_object_append(
+	vmRef *C.struct_JsonnetVm,
+	obj *C.struct_JsonnetJsonValue,
+	f *C.char,
+	v *C.struct_JsonnetJsonValue,
+) {
+	_ = getVM(vmRef) // Here we check it, is it ok?
+	d := getJSONValue(obj)
+	table, ok := d.obj.(map[string]interface{})
+
+	if !ok {
+		fmt.Fprintf(os.Stderr, "object should be provided")
+		os.Exit(1)
+	}
+
+	val := getJSONValue(v)
+	table[C.GoString(f)] = val.obj
+	d.children = append(d.children, v)
+}
+
+//export jsonnet_json_destroy
+func jsonnet_json_destroy(vmRef *C.struct_JsonnetVm, v *C.struct_JsonnetJsonValue) {
+	_ = getVM(vmRef) // Here we check it, is it ok?
+
+	for _, child := range getJSONValue(v).children {
+		// TODO what if a child is already freed?
```

Comment:
Well, double free. But the ownership of sub-values is exclusive, so the user should not let that happen (doc: https://github.com/google/jsonnet/blob/master/include/libjsonnet.h#L133, C++ impl: https://github.com/google/jsonnet/blob/master/core/libjsonnet.cpp#L149). This is acceptable behavior to me. 

BTW I would argue that the ownership there should not be exclusive, but that boat has sailed probably, and changing that is completely out of scope in this PR. 

BTW2 we could perhaps add an explicit "cleared" state, and fail with a clear message if the user saves the same JsonnetJSONValue in two places, but I'm not sure it's worth it.